display_list/geometry/dl_region.cc

```diff
@@ -0,0 +1,260 @@
+DlRegion::~DlRegion() {
+  for (auto& spanvec : spanvec_pool_) {
+    delete spanvec;
+  }
+  for (auto& line : lines_) {
+    delete line.spans;
+  }
+}
+
+void DlRegion::addRect(const SkIRect& rect) {
+  if (rect.isEmpty()) {
+    return;
+  }
+
+  int32_t i1 = rect.fTop;
+  int32_t i2 = rect.fBottom;
+
+  size_t dirty_start = -1;
+  size_t dirty_end = 1;
+
+  // Marks line as dirty. Dirty lines will be checked for equality
+  // later and merged as needed.
+  auto mark_dirty = [&](size_t line) {
+    if (dirty_start == static_cast<size_t>(-1)) {
+      dirty_start = line;
+      dirty_end = line;
+    } else {
+      dirty_start = std::min(dirty_start, line);
+      dirty_end = std::max(dirty_end, line);
+    }
+  };
+
+  auto upper_bound = std::upper_bound(
+      lines_.begin(), lines_.end(), i1,
+      [](int32_t i, const SpanLine& line) { return i < line.bottom; });
+
+  auto start_index = upper_bound - lines_.begin();
+
+  for (size_t i = start_index; i < lines_.size() && i1 < i2; ++i) {
+    SpanLine& line = lines_[i];
+
+    // If this is false the start index is wrong.
+    assert(i1 < line.bottom);
+
+    if (i2 <= line.top) {
+      insertLine(i, makeLine(i1, i2, rect.fLeft, rect.fRight));
+      mark_dirty(i);
+      i1 = i2;
+      break;
+    }
+    if (i1 < line.top) {
+      auto prevLineStart = line.top;
+      insertLine(i, makeLine(i1, prevLineStart, rect.fLeft, rect.fRight));
+      mark_dirty(i);
+      i1 = prevLineStart;
+      continue;
+    }
+    if (i1 > line.top) {
+      // duplicate line
+      auto prevLineEnd = line.bottom;
+      line.bottom = i1;
+      mark_dirty(i);
+      insertLine(i + 1, makeLine(i1, prevLineEnd, *line.spans));
+      continue;
+    }
+    assert(i1 == line.top);
+    if (i2 < line.bottom) {
+      // duplicate line
+      auto newLine = makeLine(i2, line.bottom, *line.spans);
+      line.bottom = i2;
+      line.insertSpan(rect.fLeft, rect.fRight);
+      insertLine(i + 1, newLine);
+      i1 = i2;
+      mark_dirty(i);
+      break;
+    }
+    assert(i2 >= line.bottom);
+    line.insertSpan(rect.fLeft, rect.fRight);
+    mark_dirty(i);
+    i1 = line.bottom;
+  }
+
+  if (i1 < i2) {
+    lines_.push_back(makeLine(i1, i2, rect.fLeft, rect.fRight));
+    mark_dirty(lines_.size() - 1);
+  }
+
+  // Check for duplicate lines and merge them.
+  if (dirty_start != static_cast<size_t>(-1)) {
+    // Expand the region by one if possible.
+    if (dirty_start > 0) {
+      --dirty_start;
+    }
+    if (dirty_end + 1 < lines_.size()) {
+      ++dirty_end;
+    }
+    for (auto i = lines_.begin() + dirty_start;
+         i < lines_.begin() + dirty_end;) {
+      auto& line = *i;
+      auto& next = *(i + 1);
+      if (line.bottom == next.top && line.spansEqual(next)) {
+        --dirty_end;
+        next.top = line.top;
+        i = removeLine(i);
+      } else {
+        ++i;
+      }
+    }
+  }
+}
+
+std::vector<SkIRect> DlRegion::getRects(bool deband) const {
+  std::vector<SkIRect> rects;
```

Comment:
I think the algorithm below only works as long as rects remains sorted by bottom coordinate? Is that worth mentioning here in a comment.

New rectangles are inserted at the end of the rects vector and they are only ever grown upward to absorb a previous rect so they can never violate the bottom ordering.

Just some things that I had to realize about the code before I could see if it was working as intended.